onnx/defs/schema.cc

```diff
@@ -110,11 +110,11 @@ void OpSchema::Verify(const NodeProto& node) const {
 
   // Check the values of inputs / outputs
   for (int in_idx = 0; in_idx < node.input_size(); ++in_idx) {
-    if (node.input(in_idx).empty() && !(inputs_[in_idx].IsOptional())) {
+    if (node.input(in_idx).empty() && (Single == inputs_[in_idx].GetOption())) {
```

Comment:
Do we want to permit only the last parameter to be variadic? If we allow a parameter other than the last one to be variadic (still restricting the number of variadic parameters to be one), then the association between the index of an actual parameter and the index of the corresponding formal parameter is more complex. For parameters before the optional one, it is a direct match. For parameters after the optional one, we need to count from the back ... that is, the k-th parameter from the end in the actual parameter list will correspond to the k-th parameter from the end in the formal parameter list.

Reply:
I'd say to allow variadic be at the end only, it makes it easier to reason (seems like you're already checking it below)

I also realized that this code had bug before, we need to check that `in_idx < inputs_.size()`